### ISSUE TITLE
Support weak relations with no foreign key

### ISSUE THREAD

**Issue by davesteinberg** [prisma/prisma#7351]
## Problem

I need a relation in my schema so that I can query across two tables, but without the referential integrity constraints that come with a foreign key. This is related to issue #7293, but different: I am using a database that supports foreign keys (PostgreSQL) and I want them for some relations in my schema but not for others.

## Suggested solution

This [this discussion](https://github.com/prisma/prisma/discussions/7274), @janpio showed me that I can do what I need if I just don't run prisma migrate after adding the relation to my schema, but I want to continue to evolve my schema and use prisma migrate in the future. So, what I really need is a way to tell Prisma that a particular relation should not have a foreign key.

That could be achieved by supporting a new `weak` argument on a `@relation` attribute:

```
  postalCode PostalCode? @relation(fields: [code], references: [code], weak: true)
```

But I would be open to any other alternative.

## Alternatives

Beyond not letting prisma migrate see the relation, I don't have  other ideas.

## Additional context

Here are the details of my use case, which I have copied from the discussion linked above...

Briefly, it's about using some relevant data, if available, to enrich another entity. The key thing is "if available". It's a weaker relationship in that you don't know if that relevant data will be there until you look, as opposed to a definite relationship between two entities that should/can be enforced.

My schema already has regular relationships with the need for referential integrity, but now I want to add one of these weaker relationships, as well.

This is the relevant portion of my model:

```
model Story {
  id         String      @id @default(cuid())
  createdAt  DateTime    @default(now())
  content    String
  code       String
  postalCode PostalCode? @relation(fields: [code], references: [code])
}

model PostalCode {
  code     String   @id
  province Province
  name     String
  hotspot  Boolean
}

enum Province { ... }
```

The `code` in `Story` is what defines the weak relationship between stories and postal codes. Every story will always have a `code`, but there may or may not be an actual `PostalCode` entity with that code.

I'm going to be initializing the `PostalCode` table as part of deploying the app, loading it up with data for each postal code, including an official placename from the post office, along with information from other sources, such as whether the postal code is a hotspot. Users won't ever change or add to this data. It will only be updated when I receive new data from the post office, process it, and load it.

When a user submits a story, they must type in the digits of their postal code. It may match one in the `PostalCode` table or it may not. It might not match one if it's a newly created postal code and our data isn't up to date yet or even just because they entered a fake postal code. That's totally fine.

Later, when another user visits the site, we'll retrieve the list of stories to display to them. At that point, I want to join that with the postal code data by the code that the submitter had entered. That way, I'll be able to show the story enhanced with additional information about the postal code: the placename and an indicator if it's a hotspot. If there is no matching `PostalCode` entity, no problem, I'll just show the code the submitter typed, with no name and no hotspot indicator.

Clearly this kind of query is totally possible without having a foreign key. The only thing a foreign key would do is prevent adding the story in the first place if it doesn't match a postal code. That's why a foreign key won't work in this situation.

And there's one extra complication: we already have the `Story` model (without the `postalCode` relation) in the application, which is live and running today. Currently, we just display the postal code that the submitter typed. I want to add the `PostalCode` model and data so that we can show more helpful information about the stories we have and the new stories to come. With hundreds of stories and no postal codes in the database, I can't even get as far as successfully running migrate on the schema if doing so adds the foreign key.

**Comment by janpio** [comment]
This is how this works for TypeORM: https://typeorm.io/#/relations-faq/avoid-foreign-key-constraint-creation

**Comment by kyoungduck** [comment]
I need this too.

In my case, I dont use FK in my database for following reasons:

1. FK cannot be used for partition tables.
2. FK can lock the database, so lock contention can be occurred and query latency will be increased.
3. Eventually, DBMS performance will be slowed down

number 2,3 is talking about performance but, 1 becomes a constraint cannot to use FK.

**Comment by davesteinberg** [comment]
@pimeys That looks like a very helpful enhancement for databases without foreign keys, but I don't think it addresses this issue. As I said above, "I am using a database that supports foreign keys (PostgreSQL) and I want them for some relations in my schema but not for others."

Am I misunderstanding, or isn't https://github.com/prisma/prisma-engines/pull/2212 an all-or-nothing option for the whole schema?

**Comment by pimeys** [comment]
Yup, you are right, we closed the wrong issue. The correct one is https://github.com/prisma/prisma/issues/7293

Sorry about that, switching the issues and reopening this.

**Comment by Stormtv** [comment]
This would really help me solve my issues, currently I'm using`relationMode = "prisma"` when creating the migrations and then removing `relationMode` when generating the client. Super inefficient and not a great user experience this seems like something that wouldn't be too hard to implement and would be really helpful in high throughput applications.

I think this is a more flexible solution compared to https://github.com/prisma/prisma/issues/15759

**Comment by rskvazh** [comment]
Same here. Want a weak relation because some optional 1:1 relation may be created before the main relation row. And I want to use prisma's include option on that and foreign keys for other relations.

**Comment by luan** [comment]
To add some fuel to the fire. In MySQL the engine "Memory" `ENGINE=MEMORY` doesn't support foreign keys, but otherwise works pretty well with Prisma. Without the option to generate weak relations Prisma will keep trying to add a foreign key even though MySQL just ignores it.

**Comment by chrisui** [comment]
I documented another use case (event sourcing and ephemeral projection tables) in https://github.com/prisma/prisma/issues/20890 before finding this issue. Copied here.

> Problem
Following an event sourcing pattern, imagine you wanted to build event projections which project into postgres and wanted to manage your schema and queries using prisma. When "rebuilding" a projection you must truncate a table, temporarily accepting inconsistent state, and rebuild from your stream of events. Of course this falls down when using prisma because of the default foreign key implementation.
Ie. When using these patterns we do not want foreign keys, or any emulation, as we treat the tables ephemerally.

> Suggested solution
As the only alternative to foreign keys still tries to emulate the integrity we need another mode.

> Allow a third relationMode option which a) creates indexes when defining a relation and b) does not try to "emulate" integrity - I named the request "indexed-no-integrity" as it seems pretty descriptive but could be anything
Allow this to be configured on a per relation basis (like described in https://github.com/prisma/prisma/issues/10611) - eg. in our schema we do have some models that we would want to enforce integrity of and some which we dont as they are treated ephemerally - we want to manage them all with prisma!

> Alternatives
The only workarounds right now are:
> [untested] to manually update every migration to exclude foreign keys on projection tables but do create indexes.
Model without any @relation and forsake any ergonomics or efficiency of nested querying

> Additional context
See reference for [event projections](https://event-driven.io/en/projections_and_read_models_in_event_driven_architecture/).

**Comment by janpio** [comment]
@peterHakio Yeah, I think that was wrong back in the day. But I don't think that is a thing that I can change. I updated the mention in the text, but still on both sides the issue/PR can not be de-selected under `Development`. If you tell me how to remove this, happy to. Otherwise: The issue is still open, we are aware, the "fixed by" has no influence on anything on our side.

**Comment by mdavish** [comment]
I would find this really useful as well. I have a situation where I have these two tables:

```prisma
model EmailSubscription {
  businessId   String
  emailAddress String
  createdAt    DateTime            @default(now())
  updatedAt    DateTime            @updatedAt
  state        EmailMarketingState @default(SUBSCRIBED)
  business     Business            @relation(fields: [businessId], references: [id], onDelete: Cascade)

  @@id([businessId, emailAddress])
  @@unique([businessId, emailAddress])
}

model ShopifyCustomer {
  id                            BigInt   @id @unique
  emailAddress                  String?
  rawJson                       Json?
  businessId                    String
  business                      Business @relation(fields: [businessId], references: [id], onDelete: Cascade)

  @@unique([businessId, emailAddress])
}
```

And I want to relate them based on bot h`emailAddress` and `businessId`, BUT there is no guarantee that either side of that relationship will exist, so a foreign key doesn't work. 

(For example, some people will have Email Subscriptions but no Shopify Customer, and vice-versa.)

A "weak" relationship without a foreign key would solve this for me.

**Comment by milkliker** [comment]
> I would find this really useful as well. I have a situation where I have these two tables:
> 
> ```prisma
> model EmailSubscription {
>   businessId   String
>   emailAddress String
>   createdAt    DateTime            @default(now())
>   updatedAt    DateTime            @updatedAt
>   state        EmailMarketingState @default(SUBSCRIBED)
>   business     Business            @relation(fields: [businessId], references: [id], onDelete: Cascade)
> 
>   @@id([businessId, emailAddress])
>   @@unique([businessId, emailAddress])
> }
> 
> model ShopifyCustomer {
>   id                            BigInt   @id @unique
>   emailAddress                  String?
>   rawJson                       Json?
>   businessId                    String
>   business                      Business @relation(fields: [businessId], references: [id], onDelete: Cascade)
> 
>   @@unique([businessId, emailAddress])
> }
> ```
> 
> And I want to relate them based on bot h`emailAddress` and `businessId`, BUT there is no guarantee that either side of that relationship will exist, so a foreign key doesn't work.
> 
> (For example, some people will have Email Subscriptions but no Shopify Customer, and vice-versa.)
> 
> A "weak" relationship without a foreign key would solve this for me.

Same needs like this. A "weak" relationship would be very useful.

**Comment by fred-adept** [comment]
One potential workaround for some usecases is to create a view of the join and use that https://www.prisma.io/docs/orm/prisma-schema/data-model/views

**Comment by wilfredjonathanjames** [comment]
I'm currently syncing a few dozen models from a handful of service providers. Millions of records. I don't need hard foreign key constraints and the lack of weak relations has dramatically increased the complexity and latency of the available solutions. To maintain the ergonomics of nested queries I'm having to hold onto objects in queue until all foreign key constraints are satisfied. It's not only more complex, it's significantly slower and is wasting compute.

I for one vote for prioritising this ticket 🔝

**Comment by tuan-nguyen-wareflex** [comment]
any update? I have some issue and I want weak relation.

**Comment by rcbevans** [comment]
I have a few situations where this would be incredibly useful.  In my datamodel I have both a "User" (of my service) and a "Contact" representing a person in my knowledge graph.  Both are keyed on emailAddress, but a User may not exist where a Contact does, or vice versa.  Similarly, a Contact may exist where a CRMContact/Lead does not (or vice versa).  Right now I have to make a second round of queries after collecting a list of contacts to optionally match them with a user, if the user exists.  It would be incredibly useful to have an unconstrained (weak) relation which can pull a matching object based on the join criteria IFF it exists without a fk constraint.

**Comment by elias-summermatter** [comment]
We would need this one too

**Comment by Joaquim-Ferrer** [comment]
This would be incredibly helpful for me too. There's no good workaround.

**Comment by flawnn** [comment]
I can +1 this - this would be great!

**Comment by np42** [comment]
I needs weak relations for stepped refactoring purpose, my workaround for now is kysely

**Comment by mklimkin0190** [comment]
Absolutely need this.

**Comment by aacoimbra** [comment]
Would like to have this feature too.

**Comment by ericvtheg** [comment]
This would be incredible for me as well

**Comment by sp1r1t** [comment]
I also have a strong usecase for this. I'm loading referenced data through various API endpoints. It is not possible to tell in advanced if the reference is loaded already. If it is not present the include should just be `null`. This behaviour can be achieved by manually removing the foreign key constraints, but `migrate dev` keeps readding them.

I would love to see weak foreign key constraints when a reference is marked with `?`. This would make so much sense.

**Comment by dfasoro** [comment]
I also need this as well. My use-case includes not wanting an FK cascade for API performance predictability, DB performance/locking etc reasonings. Also, I'd like to be able to perform queries on tables across multiple schemas, or partitions in postgres. I can tolerate eventual consistency and can build async cleanup workflows.

Has anyone looked at these combination?
1. Use [prisma relation emulation mode](https://www.prisma.io/docs/orm/prisma-schema/data-model/relations/relation-mode#emulate-relations-in-prisma-orm-with-the-prisma-relation-mode), so the relations are not created within the Migration DDLs
2. A [prisma client extension](https://www.prisma.io/docs/orm/prisma-client/client-extensions) to intercept the referential actions emulation within prisma and make it a NO-OP.

**Comment by mel-habip** [comment]
Hello people 👋 have there been any updates on this? I +1 this feature too, particularly with the relation-specific `weak: true` kind of implementation.

**Comment by madlep** [comment]
Another use case where this is causing issues is to do with eventual consistency with different processes writing to different postgresql tables.

We've got a parent model that looks like:
```prisma
model Product {
  id     String @id @default(cuid())
  images Image[]
  //  ...
}
```

And child images attached:
```prisma
model Image {
  id           String @id @default(cuid())
  product_id   String
  product      Product @relation(fields: [product_id], references: [id])
  original_url String
  cdn_url      String
  // ..
}
```

Products are created by users in a separate service - including a generated id. When products are ready to be published, JSON payload about the product is published to a pubsub topic, and separate handlers execute for creating the `Product` (including a lot of other business logic), and handlers for pushing the product images to a CDN. We don't want these processes coupled, or have to set everything up so that one is temporally ordered to execute first.

If `Image` rows are created before their parent `Product`, it is fine. We aren't querying those, and the data will be eventually consistent. In the mean time, we're perfectly happy with temporary missing parent products for an image.

But we want to still be able to reference `images` when querying `Product`, however Prisma creates the unwanted foreign key constraint whenever migrations are generated - which isn't appropriate for this use case.

**Comment by SarenT** [comment]
I want to add another use case here. I have a DB model, which is based on some specifications. Sometimes, we receive data in this format, which does not follow the specs and fail at insertions. Loose relationships without foreign key enforcement would benefit a lot, without fiddling with SQL migrations.

**Comment by mdesousa** [comment]
this would be very useful, see #28505

**Comment by zacaj** [comment]
Feel like I run into this every 3-6 months and have to redesign some feature once I remember that prisma doesn't support this

**Comment by JimGitFE** [comment]
## A Quick Solution: **Weak Relationships**

Note, `upsert` should be the solution for all cases (as designed by [prisma](https://www.prisma.io/docs/orm/prisma-client/queries/crud#update-or-create-records)). Either way this is the method I use to remove **Foreign Key Constrains**:

1. Add `?` Optional Foreign Key at `schema.prisma`
```prisma
feKey String? @unique
user User? @relation(fields: [feKey], references: [id])
```

2. `npx prisma migrate dev --name weak_relationship --create-only` (**Important**: --create-only)
```sql
-- file: /migrations/<id>_weak_relationship/migration.sql

-- DropForeignKey
ALTER TABLE "Visit" DROP CONSTRAINT "Visit_feKey_fkey";

-- AlterTable
ALTER TABLE "Visit" ALTER COLUMN "feKey" DROP NOT NULL;

-- AddForeignKey
ALTER TABLE "Visit" ADD CONSTRAINT "Visit_feKey_fkey" FOREIGN KEY ("feKey") REFERENCES "User"("id") ON DELETE SET NULL ON UPDATE CASCADE;
```

3. Remove SQL Constraint at `/migrations/<id>_weak_relationship/migration.sql`
```sql
-- file: migration.sql

-- DropForeignKey
ALTER TABLE "Visit" DROP CONSTRAINT "Visit_feKey_fkey";

-- AlterTable
ALTER TABLE "Visit" ALTER COLUMN "feKey" DROP NOT NULL;
```
4. `npx prisma db push` (forces changes) or `npx prisma migrate dev` (prisma might re-migrate with missing `ADD CONSTRAINT`)
5. `npx prisma generate` (Client now ready)

So while losing **Referential Integrity** at the Database level, we keep an **Implicit Relation** at the Client level. Hope it helps someone!
